ToOne relationships that are marked as FetchType.LAZY cannot be cast to their subtypes without a manual refresh

Let's say we have this example:

```
public class MySku extends SkuImpl
    @Column("NEW_PROPERTY")
     public String myProperty;
     ...

    public String getMyProperty() {
        return myProperty;
    }
}
```

Then ensure that you create a `MySku` instance in the database and then add that to your cart. In order to get around the problem of a ton of joins with catalog and order extensions (see #1209), we have changed the relationship from `DiscreteOrderItem` to `Sku` to be `FetchType.LAZY`:

```
@ManyToOne(fetch = FetchType.LAZY, targetEntity = SkuImpl.class, optional=false)
@JoinColumn(name = "SKU_ID", nullable = false)
@Index(name="DISCRETE_SKU_INDEX", columnNames={"SKU_ID"})
@AdminPresentation(friendlyName = "DiscreteOrderItemImpl_Sku", order=Presentation.FieldOrder.SKU,
        group = OrderItemImpl.Presentation.Group.Name.Catalog, groupOrder = OrderItemImpl.Presentation.Group.Order.Catalog)
@AdminPresentationToOneLookup()
protected Sku sku;
```

So, when you call `discreteOrderItem.getSku()`, the instance that you actually get back is a javaassist object (aka a Hibernate proxy) that conforms to SkuImpl (it just looks at the `@ManyToOne` information to determine this). Thus, this invocation will fail with a class cast exception:

```
// Throws ClassCastException
String myProperty = ((MySku) discreteOrderItem.getSku()).getMyProperty();
```

Casting the `getSku()` call to a `Sku` or `SkuImpl` will always succeed.

Here's where it gets fun. Because it's loaded lazily, when you try to actually get a property from the Sku it goes and looks it up (db or cache) and then gives you the right type. Thus, if you do this then it will succeed:

```
// force a load of the Sku
discreteOrderItem.getSku().getId();
String myProperty = ((MySku) discreteOrderItem.getSku()).getMyProperty();
// all good
```

The above is a workaround. What we should do in the framework is in the getter, go ahead and refresh the Sku instance. According to the almight internet at http://stackoverflow.com/questions/17155452/hibernate-lazy-loading-for-reverse-one-to-one-workaround-how-does-this-work, it looks like the addition of `@LazyToOne(LazyToOneOption.NO_PROXY)` to that property should work because then whenever the reference is accessed the instance will be refreshed automatically.

The DiscreteOrderItem -> Sku relationship is 1 instance but I'm sure there are others. Doing a quick search in my projects:

- BundleOrderItem - sku and product
- DiscreteOrderItem - sku and product
- OrderItem - category and giftWrapOrderItem

Credit to @pbaggett for going through the debugging to figure this out.

@jefffischer you might have an opinion here.

In an effort to get `@LazyToOne(LazyToOneOption.NO_PROXY)` working for us I attempted to use the hibernate build-time bytecode enhancement but ran into the following exception on many fields during build:
```
[ERROR] Unable to enhance class [org.broadleafcommerce.core.rating.domain.ReviewDetailImpl]
org.hibernate.bytecode.enhance.spi.EnhancementException: Unable to enhance persistent attribute [org.broadleafcommerce.core.rating.domain.ReviewDetailImpl:customer]
	at org.hibernate.bytecode.enhance.internal.PersistentAttributesEnhancer.enhancePersistentAttribute(PersistentAttributesEnhancer.java:104)
	at org.hibernate.bytecode.enhance.internal.PersistentAttributesEnhancer.enhance(PersistentAttributesEnhancer.java:69)
	at org.hibernate.bytecode.enhance.internal.EntityEnhancer.enhance(EntityEnhancer.java:71)
	at org.hibernate.bytecode.enhance.spi.Enhancer.enhance(Enhancer.java:161)
	at org.hibernate.bytecode.enhance.spi.Enhancer.enhance(Enhancer.java:107)
	at org.hibernate.bytecode.enhance.plugins.MavenEnhancePlugin.processEntityClassFile(MavenEnhancePlugin.java:150)
.......
Caused by: javassist.NotFoundException: org.broadleafcommerce.profile.core.domain.Customer
	at javassist.ClassPool.get(ClassPool.java:450)
	at javassist.bytecode.Descriptor.toCtClass(Descriptor.java:592)
	at javassist.bytecode.Descriptor.toCtClass(Descriptor.java:542)
	at javassist.CtField.getType(CtField.java:371)
	at org.hibernate.bytecode.enhance.internal.AttributeTypeDescriptor.resolve(AttributeTypeDescriptor.java:102)
```
I then tried activating the hibernate runtime bytecode weaver by adding `<property name="hibernate.ejb.use_class_enhancer" value="true"/>` to persistence-common.xml which resulted in the following exception frequently, and the `@LazyToOne(LazyToOneOption.NO_PROXY)` still had no effect:
```
[artifact:mvn] [ WARN] 16:29:41 ClassFileTransformerAdapter - Error weaving class [null] with transformer of class [org.hibernate.ejb.instrument.InterceptFieldClassFileTransformer]
[artifact:mvn] java.lang.instrument.IllegalClassFormatException
[artifact:mvn] 	at org.hibernate.ejb.instrument.InterceptFieldClassFileTransformer.transform(InterceptFieldClassFileTransformer.java:79)
[artifact:mvn] 	at org.springframework.orm.jpa.persistenceunit.ClassFileTransformerAdapter.transform(ClassFileTransformerAdapter.java:56)
[artifact:mvn] 	at org.springframework.instrument.classloading.InstrumentationLoadTimeWeaver$FilteringClassFileTransformer.transform(InstrumentationLoadTimeWeaver.java:184)
[artifact:mvn] 	at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
[artifact:mvn] 	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
......
```
This is caused by lambda expressions in java 8 because they have no class name, there is a bug open for this currently: https://hibernate.atlassian.net/browse/HHH-9541

I will now simply add a `deproxy()` method that can be called where we use lazy loading to receive the actual concrete instance.

--- a/common/src/main/java/org/broadleafcommerce/common/util/HibernateUtils.java
+++ b/common/src/main/java/org/broadleafcommerce/common/util/HibernateUtils.java
@@ -1,0 +1,47 @@
+/*
+ * #%L
+ * BroadleafCommerce Common Libraries
+ * %%
+ * Copyright (C) 2009 - 2013 Broadleaf Commerce
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package org.broadleafcommerce.common.util;
+
+import org.hibernate.proxy.HibernateProxy;
+import org.hibernate.proxy.LazyInitializer;
+
+/**
+ * Hibernate convenience methods
+ *
+ * @author Philip Baggett (pbaggett)
+ */
+public class HibernateUtils {
+    /**
+     * <p>Ensure a domain object is an actual persisted object and not a Hibernate proxy object by getting its real implementation.
+     *
+     * <p>This is primarily useful when retrieving a lazy loaded object that has been subclassed and you have the intention of casting it.
+     *
+     * @param t the domain object to deproxy
+     * @return the actual persisted object or the passed in object if it is not a Hibernate proxy
+     */
+    public static <T> T deproxy(T t) {
+        if (t instanceof HibernateProxy) {
+            HibernateProxy proxy = (HibernateProxy)t;
+            LazyInitializer lazyInitializer = proxy.getHibernateLazyInitializer();
+            return (T)lazyInitializer.getImplementation();
+        }
+        return t;
+    }
+}

--- a/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/BundleOrderItemImpl.java
+++ b/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/BundleOrderItemImpl.java
@@ -26,6 +26,7 @@ import org.broadleafcommerce.common.presentation.AdminPresentationClass;
 import org.broadleafcommerce.common.presentation.AdminPresentationCollection;
 import org.broadleafcommerce.common.presentation.AdminPresentationToOneLookup;
 import org.broadleafcommerce.common.presentation.client.SupportedFieldType;
+import org.broadleafcommerce.common.util.HibernateUtils;
 import org.broadleafcommerce.core.catalog.domain.Product;
 import org.broadleafcommerce.core.catalog.domain.ProductBundle;
 import org.broadleafcommerce.core.catalog.domain.ProductBundleImpl;
@@ -102,7 +103,7 @@ public class BundleOrderItemImpl extends OrderItemImpl implements BundleOrderIte
 
     @Override
     public Sku getSku() {
-           return sku;
+        return HibernateUtils.deproxy(sku);
     }
 
     @Override
@@ -122,7 +123,7 @@ public class BundleOrderItemImpl extends OrderItemImpl implements BundleOrderIte
 
     @Override
     public ProductBundle getProductBundle() {
-        return productBundle;
+        return HibernateUtils.deproxy(productBundle);
     }
 
     @Override

--- a/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/DiscreteOrderItemImpl.java
+++ b/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/DiscreteOrderItemImpl.java
@@ -27,6 +27,7 @@ import org.broadleafcommerce.common.presentation.AdminPresentation;
 import org.broadleafcommerce.common.presentation.AdminPresentationClass;
 import org.broadleafcommerce.common.presentation.AdminPresentationToOneLookup;
 import org.broadleafcommerce.common.presentation.client.SupportedFieldType;
+import org.broadleafcommerce.common.util.HibernateUtils;
 import org.broadleafcommerce.core.catalog.domain.Product;
 import org.broadleafcommerce.core.catalog.domain.ProductImpl;
 import org.broadleafcommerce.core.catalog.domain.Sku;
@@ -121,7 +122,7 @@ public class DiscreteOrderItemImpl extends OrderItemImpl implements DiscreteOrde
 
     @Override
     public Sku getSku() {
-        return sku;
+        return HibernateUtils.deproxy(sku);
     }
 
     @Override
@@ -144,7 +145,7 @@ public class DiscreteOrderItemImpl extends OrderItemImpl implements DiscreteOrde
 
     @Override
     public Product getProduct() {
-        return product;
+        return HibernateUtils.deproxy(product);
     }
 
     @Override

--- a/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/OrderItemImpl.java
+++ b/core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/order/domain/OrderItemImpl.java
@@ -39,6 +39,7 @@ import org.broadleafcommerce.common.presentation.override.AdminPresentationMerge
 import org.broadleafcommerce.common.presentation.override.AdminPresentationMergeOverride;
 import org.broadleafcommerce.common.presentation.override.AdminPresentationMergeOverrides;
 import org.broadleafcommerce.common.presentation.override.PropertyType;
+import org.broadleafcommerce.common.util.HibernateUtils;
 import org.broadleafcommerce.core.catalog.domain.Category;
 import org.broadleafcommerce.core.catalog.domain.CategoryImpl;
 import org.broadleafcommerce.core.offer.domain.CandidateItemOffer;
@@ -293,7 +294,7 @@ public class OrderItemImpl implements OrderItem, Cloneable, AdminMainEntity, Cur
 
     @Override
     public Category getCategory() {
-        return category;
+        return HibernateUtils.deproxy(category);
     }
 
     @Override
@@ -394,7 +395,7 @@ public class OrderItemImpl implements OrderItem, Cloneable, AdminMainEntity, Cur
 
     @Override
     public GiftWrapOrderItem getGiftWrapOrderItem() {
-        return giftWrapOrderItem;
+        return HibernateUtils.deproxy(giftWrapOrderItem);
     }
 
     @Override

